node/core/pvf/common/src/worker.rs

```diff
@@ -174,20 +181,26 @@ pub fn stringify_panic_payload(payload: Box<dyn Any + Send + 'static>) -> String
 	}
 }
 
-/// In case of node and worker version mismatch (as a result of in-place upgrade), send `SIGTERM`
-/// to the node to tear it down and prevent it from raising disputes on valid candidates. Node
-/// restart should be handled by the node owner. As node exits, unix sockets opened to workers
-/// get closed by the OS and other workers receive error on socket read and also exit. Preparation
-/// jobs are written to the temporary files that are renamed to real artifacts on the node side, so
-/// no leftover artifacts are possible.
-fn kill_parent_node_in_emergency() {
+/// In case of node and worker version mismatch (as a result of in-place upgrade or
+/// incorrectly-built binaries), send `SIGTERM` to the node to tear it down and prevent it from
+/// raising disputes on valid candidates. Node restart should be handled by the node owner. As node
+/// exits, unix sockets opened to workers get closed by the OS and other workers receive error on
+/// socket read and also exit. Preparation jobs are written to the temporary files that are renamed
+/// to real artifacts on the node side, so no leftover artifacts are possible.
+fn kill_parent_node_in_emergency(worker_pid: u32) {
 	unsafe {
 		// SAFETY: `getpid()` never fails but may return "no-parent" (0) or "parent-init" (1) in
 		// some corner cases, which is checked. `kill()` never fails.
 		let ppid = libc::getppid();
 		if ppid > 1 {
 			libc::kill(ppid, libc::SIGTERM);
+		} else {
+			gum::error!(target: LOG_TARGET, %worker_pid, "unexpected ppid {}", ppid);
 		}
+
+		// Explicitly terminate the worker here. Not strictly necessary as it would end when it
+		// fails to read from the socket later.
+		libc::kill(worker_pid as i32, libc::SIGKILL);
```

Comment:
Then why do it? Not saying we should not, but there should be some reason.

Reply:
The idea was just to be explicit, and so we don't rely on coincidental architectural details that may change.

But, maybe we should instead have it be explicit that the worker dies when it can't read from the socket. The parent may also die for some reason other than us killing it.

Reply:
I would say, if we already die when read fails, let's leave it at that, but document that error handling accordingly.